File: sdk/src/main/java/com/libreshockwave/cast/BitmapInfo.java
package com.libreshockwave.cast;

import com.libreshockwave.io.BinaryReader;

import java.nio.ByteOrder;

/**
 * Bitmap-specific cast member data.
 * Contains dimensions, bit depth, palette, and registration point.
 */
public record BitmapInfo(
    int width,
    int height,
    int regX,
    int regY,
    int bitDepth,
    int paletteId
) {

    public static BitmapInfo parse(byte[] data) {
        if (data == null || data.length < 20) {
            return new BitmapInfo(0, 0, 0, 0, 1, 0);
        }

        BinaryReader reader = new BinaryReader(data, ByteOrder.BIG_ENDIAN);

        reader.skip(2); // unknown bytes
        reader.skip(4); // unknown

        int height = reader.readU16();
        int width = reader.readU16();

        reader.skip(8); // unknown (4 x u16)

        int regY = reader.readI16();
        int regX = reader.readI16();

        reader.skip(1); // unknown

        int bitDepth = 1;
        int paletteId = 0;

        if (reader.bytesLeft() > 0) {
            bitDepth = reader.readU8();
            reader.skip(2); // unknown
            if (reader.bytesLeft() >= 2) {
                paletteId = reader.readI16() - 1;
            }
        }

        return new BitmapInfo(width, height, regX, regY, bitDepth, paletteId);
    }

    public int bytesPerPixel() {
        return switch (bitDepth) {
            case 1 -> 0; // 1-bit packed
            case 2 -> 0; // 2-bit packed
            case 4 -> 0; // 4-bit packed
            case 8 -> 1;
            case 16 -> 2;
            case 24 -> 3;
            case 32 -> 4;
            default -> 1;
        };
    }

    public boolean isPaletted() {
        return bitDepth <= 8;
    }
}

File: cpp/src/cast/BitmapInfo.cpp
#include "libreshockwave/cast/BitmapInfo.hpp"

#include "libreshockwave/io/BinaryReader.hpp"

namespace libreshockwave::cast {

int BitmapInfo::regXLocal() const { return regX - initialLeft; }
int BitmapInfo::regYLocal() const { return regY - initialTop; }

int BitmapInfo::bytesPerPixel() const {
    switch (bitDepth) {
        case 1:
        case 2:
        case 4:
            return 0;
        case 8:
            return 1;
        case 16:
            return 2;
        case 24:
            return 3;
        case 32:
            return 4;
        default:
            return 1;
    }
}

bool BitmapInfo::isPaletted() const {
    return bitDepth <= 8;
}

BitmapInfo BitmapInfo::parse(const std::vector<std::uint8_t>& data, int directorVersion) {
    if (data.size() < 10) {
        return BitmapInfo{0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, false};
    }

    io::BinaryReader reader(data, io::ByteOrder::BigEndian);
    const int rawPitch = reader.readU16();
    const int top = reader.bytesLeft() >= 2 ? reader.readI16() : 0;
    const int left = reader.bytesLeft() >= 2 ? reader.readI16() : 0;
    const int bottom = reader.bytesLeft() >= 2 ? reader.readI16() : 0;
    const int right = reader.bytesLeft() >= 2 ? reader.readI16() : 0;
    const int height = bottom - top;
    const int width = right - left;

    int regX = 0;
    int regY = 0;
    int bitDepth = 1;
    int paletteId = 0;
    int paletteCastLib = 0;
    int pitch = 0;
    int alphaThreshold = 0;
    bool useAlpha = false;

    if (directorVersion < 1200) {
        if (reader.bytesLeft() >= 8) reader.skip(8);
        if (reader.bytesLeft() >= 2) regY = reader.readI16();
        if (reader.bytesLeft() >= 2) regX = reader.readI16();
        if (reader.bytesLeft() >= 1) reader.skip(1);
        if (reader.bytesLeft() >= 1) {
            bitDepth = reader.readU8();
            if (directorVersion >= 1100 && reader.bytesLeft() >= 2) {
                paletteCastLib = reader.readI16();
            }
            if (reader.bytesLeft() >= 2) {
                paletteId = reader.readI16() - 1;
            }
        }
        pitch = rawPitch & 0x0FFF;
        if (bitDepth == 0) bitDepth = 1;
    } else {
        if (reader.bytesLeft() >= 1) {
            alphaThreshold = reader.readU8();
        }
        if (reader.bytesLeft() >= 7) reader.skip(7);
        if (reader.bytesLeft() >= 2) regY = reader.readI16();
        if (reader.bytesLeft() >= 2) regX = reader.readI16();
        if (reader.bytesLeft() >= 1) {
            const int updateFlags = reader.readU8();
            useAlpha = (updateFlags & 0x10) != 0;
        }
        if ((rawPitch & 0x8000) != 0) {
            pitch = rawPitch & 0x3FFF;
            if (reader.bytesLeft() >= 1) bitDepth = reader.readU8();
            if (reader.bytesLeft() >= 2) paletteCastLib = reader.readI16();
            if (reader.bytesLeft() >= 2) paletteId = reader.readI16() - 1;
        } else {
            bitDepth = 1;
            pitch = rawPitch & 0x3FFF;
        }
    }

    return BitmapInfo{width, height, regX, regY, left, top, bitDepth, paletteId, paletteCastLib, pitch, alphaThreshold, useAlpha};
}

} // namespace libreshockwave::cast
